rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,131 @@
+tag to be signed.
+- It is unable to verify container image signature by digests.
+- Signature relocation is not possible with this tool.
+- It is a project that currently has low activity and community involvement,
+  since the community is currently focused on the development of Notary v2.
+
+[`cosign`](https://github.com/sigstore/cosign) presents itself as a solution to
+some of those aforementioned challenges. With `cosign`, the user is able to:
+- Relocate container images across registries along with its signatures.
+  - Currently, there is no built-in command for relocation on `cosign`, but
+    copying container images and signatures across registries using other
+    container CLI tools such as `crane` and then verifying with `cosign` is
+    possible.
+  - There is also a proposal for
+    [implementing a `cosign cp`](https://github.com/sigstore/cosign/issues/303)
+    command that would affect this copy as a single operation.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Create a new configuration in the `Image` resource that allow users to
+  configure image signing with `cosign` as the signing mechanism, specifying
+  the secret reference to get the private key and passphrase. The user can also
+  specify signature annotations as part of configuration:
+```yaml
+cosign:
+    secretRef: <secret-name> # private key and passphrase to sign images
+    annotations: # optional map of key-value pairs the user wants to add to images
+      key1: value1
+      key2: value2
+```
+`cosign` will apply this optional map of annotations to the signature payload,
+which can then be verified with `cosign verify`. Verification is out of scope
+for this RFC.
+
+- Implement a flow that generates the signature payload for the image, then
+  calculates its signature and pushes it to the registry where the image is
+  located, using the same credentials that were used to push the image. This
+  flow must happen after the image has been pushed to the registry.
+
+- If `cosign` fails to sign an image, the build should fail and output an error
+  message specifying the failure.
+
+- Whenever `kpack` signs an image using `cosign`, it should add these
+  annotations:
+  - Build number.
+  - Cluster builder used.
+  - Buildpacks involved in the build process.
```

Comment:
The buildpacks/stack images used is available on cnb built image labels. The digest of the image will encompass these labels and the digest will be signed by cosign, is is necessary to also include this information in the cosign metadata?

Reply:
cc' @samj1912

Reply:
I guess we could skip these parts and instead extract them from the image.

Reply:
Would it be OK if I edited this point about metadata to leave only build number and build time?

Reply:
I updated this section to leave only build number and build time as required metadata.